design/BufferOperations.md

```diff
@@ -0,0 +1,148 @@
+
+The mapping operations for `WebGPUBuffer` are:
+
+```
+partial interface WebGPUBuffer {
+    WebGPUMappedMemory mapWrite(u32 offset, u32 size);
+    WebGPUMappedMemory mapRead(u32 offset, u32 size);
+};
+```
+
+These operations return new `WebGPUMappedMemory` objects representing the current range of the buffer for writing or mapping.
+The results are initialized in the "pending" state and transition at Javascript task boundary to the "available" state when the implementation can determine the GPU is done using the buffer.
+Calling `mapRead` or `mapWrite` puts the buffer in the mapped state.
+No operations are allowed in a buffer in that state except additional calls to `mapRead` or `mapWrite` and calls to `unmap`.
+In particular a mapped buffer cannot be used in a `WebGPUCommandBuffer` given to `WebGPUQueue.submit`.
+The following must be true or a validation error occurs for `mapWrite` (resp. `mapRead`):
+
+ - The buffer must have been created with the `WebGPUBufferUsage.MAP_WRITE` (resp. `WebGPUBufferUsage.MAP_READ`) usage.
+ - `offset + size` must not overflow and be at most the size of the buffer
+ - Depending on the design of memory barriers, the buffer must be, or allowed to be in the `WebGPUBufferUsage.MAP_WRITE` (resp. `WebGPUBufferUsage.MAP_READ`) usage.
+
+Then a mapped buffer can be unmapped with:
+
+```
+partial interface WebGPUBuffer {
+    void unmap();
+};
+```
+
+This operation invalidates all the `WebGPUMappedMemory` created from the buffer and puts the buffer in the unmapped state.
+The buffer must be in the mapped state otherwise a validation error occurs when `unmap` is called.
+
+### WebGPUMappedMemory
+
+`WebGPUMappedMemory` is an object representing a mapped region of a buffer that's both pollable and promise-like.
+
+It can be in one of three states: pending, available and invalidated.
+
+The pollable interface is:
+
+```
+partial interface WebGPUMappedMemory {
+    bool isPending();
+    ArrayBuffer getPointer();
+};
+```
+
+`isPending` return true if the object is in the pending state, false otherwise.
+`getPointer` returns an ArrayBuffer representing the buffer data if the object is in the available state, null otherwise.
+
+`WebGPUMappedMemory` is also `then`-able, meaning that it acts like a Javascript `Promise`:
+
+```
+partial interface WebGPUMappedMemory {
+    Promise then(WebGPUMappedMemorySuccessCallback success,
+                 optional WebGPUMappedMemoryErrorCallback error);
+};
+```
+
+This acts like a `Promise<ArrayBuffer>.then` that is resolved on the Javascript task boundary in which the implementation detects the GPU is done with the buffer.
+On that boundary:
+
+ - The `WebGPUMappedMemory` goes in the available state.
+ - If the `WebGPUMappedMemory` was created via `WebGPUBuffer.mapWrite`, its content is cleared to 0.
+ - `success` is called with the content of the memory as an argument.
```

Comment:
What is the rationale for having the Promise give you an ArrayBuffer in the writing (mapWrite or setSubData) case? When would I want my own data given back to me?

Reply:
The `mapWrite` ArrayBuffer is where you write the data and it is made available to the buffer when `unmap` is called. I'll edit the doc.

Reply:
Apologies if I didn't make myself clear in what I was asking.  

Suppose I call mapWrite, .then the promise and return control back to Javascript.  When the promise resolves, what am I meant to do with the ArrayBuffer I get in the promise?  Will that be the same ArrayBuffer as the one returned from `WebGPUMappedMemory.getPointer`?